File: dotnet/Allmark/Render/ConsoleStrikethroughRenderer.cs
namespace Allmark.Render;

using Allmark.Types;

public static class ConsoleStrikethroughRenderer
{
    public static Renderer Create(string style)
    {
        return new Renderer
        {
            Name = "strikethrough",
            Render = (node, state, first, last, decode) => Render(node, state, style),
        };
    }

    public static void Render(MarkdownNode node, RendererState state, string style)
    {
        state.Output.Append($"{style}{RenderToConsole.AnsiStrikethrough}");
        RenderChildren.Execute(node, state);
        state.Output.Append($"{RenderToConsole.AnsiStrikethroughReset}{RenderToConsole.AnsiReset}");
    }
}

File: zig/src/render-console/consoleStrikethroughRenderer.zig
const std = @import("std");

const MarkdownNode = @import("../types/MarkdownNode.zig").MarkdownNode;
const ConsoleRendererState = @import("../types/RendererState.zig").RendererState;
const Renderer = @import("../types/Renderer.zig").Renderer;
const ansiDim = @import("console.zig").ansiDim;
const ansiStrikethrough = @import("console.zig").ansiStrikethrough;
const ansiStrikethroughReset = @import("console.zig").ansiStrikethroughReset;
const ansiReset = @import("console.zig").ansiReset;
const renderChildrenConsole = @import("console.zig").renderChildrenConsole;

pub const consoleStrikethroughRenderer = Renderer{
    .name = "strikethrough",
    .render = render,
};

pub fn render(node: *const MarkdownNode, state: *ConsoleRendererState, first: ?bool, last: ?bool, decode: ?bool) void {
    _ = first;
    _ = last;
    _ = decode;

    state.output.appendSlice(state.allocator, ansiDim) catch unreachable;
    state.output.appendSlice(state.allocator, ansiStrikethrough) catch unreachable;
    renderChildrenConsole(node, state, true) catch unreachable;
    state.output.appendSlice(state.allocator, ansiStrikethroughReset) catch unreachable;
    state.output.appendSlice(state.allocator, ansiReset) catch unreachable;
}
